### PR TITLE
Implement StrongSemilatticeOfSemigroups and related things

### PR DESCRIPTION
**Pull request by MTWhyte**
This pull request is no longer a work in progress.

### CODE REVIEW

**Comment by tomcontileslie** [comment]
With the new changes the Strong Semilattice of Semigroups feature is ready to be integrated into the main branch.

**Comment by tomcontileslie** [comment]
An additional feature we could add to this branch is the following. The **Digraphs** documentation 5.5-2 has `AsSemigroup(filt, Y, gps, homs)` where:

- `filt` is necessarily `IsPartialPermSemigroup`;
- `Y` is a join- or meet- semilattice digraph;
- `gps` is a list of groups, one for each vertex;
- `homs` is a list of homomorphisms along edges in the transitive reduction of `Y`.

With our new SSS features, we would be able to add a second method working as above, but with `filt = IsStrongSemilatticeOfSemigroups`, which returns an `SSS` object.

File: doc/semicons.xml [hunk 1]
```diff
@@ -299,3 +299,174 @@ true]]></Example>
     </Description>
   </ManSection>
 <#/GAPDoc>
+
+<#GAPDoc Label="StrongSemilatticeOfSemigroups">
+  <ManSection>
+    <Func Name="StrongSemilatticeOfSemigroups" Arg="D, L, H"/>
```
Comment (wilfwilson) [inline doc/semicons.xml:305]: This is an operation, not a global function, so it should be labelled/documented as such:
```suggestion
    <Oper Name="StrongSemilatticeOfSemigroups" Arg="D, L, H"/>
```
The analogous thing applies for many of your other bits of documentation.

File: doc/semicons.xml [hunk 2]
```diff
@@ -299,3 +299,174 @@ true]]></Example>
...
+      creates the homomorphism <M>f_{41}</M> by composing the mappings along
+      paths that lead from 4 to 1, and checks that composing along all possible
+      paths produces the same result.
+    </Description>
+  </ManSection>
+<#/GAPDoc>
+
+<#GAPDoc Label="IsStrongSemilatticeOfSemigroups">
+  <ManSection>
+    <Func Name="IsStrongSemilatticeOfSemigroups" Arg="obj"/>
```
Comment (wilfwilson) [inline doc/semicons.xml:349]:
```suggestion
    <Filt Name="IsStrongSemilatticeOfSemigroups" Arg="obj"/>
```

File: tst/standard/semicons.tst [hunk 3]
```diff
@@ -949,9 +949,246 @@ gap> S := BrandtSemigroup(IsIntegerMatrixSemigroup, DihedralGroup(4), 3);
...
+gap> MultiplicativeZero(sl);
+SSSE(1, Transformation( [ 1, 1, 1 ] ))
+gap> MultiplicativeNeutralElement(sl);
+fail
+
 #
 gap> Unbind(S);
 
+# Unbind strong semilattice vars
+# TODO use less global variables
```
Comment (wilfwilson) [inline tst/standard/semicons.tst:1155]: Are you planning to actually use fewer variables? If so, please do!

File: gap/semigroups/semicons.gi [hunk 4]
```diff
@@ -805,3 +805,227 @@ for _IsXSemigroup in ["IsTransformationSemigroup",
...
+  return Concatenation("SSSE(", ViewString(x![2]), ", ", ViewString(x![3]), ")");
+end);
+
+InstallMethod(StrongSemilatticeOfSemigroups, "for a SSSE rep",
+[IsSSSERep],
+function(x)
+  return x![1];
+end);
+
+# TODO hash function currently unused
```
Comment (wilfwilson) [inline gap/semigroups/semicons.gi:1019]: Does the TODO here mean that you intend to use this?

**Review by wilfwilson — COMMENTED** [review]
Just a few quick remarks to start. That's all I've got time for at the moment.

File: gap/semigroups/semicons.gi (hunk 4 above)
Comment (tomcontileslie) [reply gap/semigroups/semicons.gi:1019]: I think we found a way around it, had initially planned to use for equality checking if I remember correctly? Can't think of a use for it now so will remove.

File: gap/semigroups/semicons.gi [hunk 5]
```diff
@@ -805,3 +805,227 @@ for _IsXSemigroup in ["IsTransformationSemigroup",
...
 Unbind(_IsXSemigroup);
+
+InstallMethod(StrongSemilatticeOfSemigroups,
+"for a digraph, a list, and a list",
+[IsDigraph, IsList, IsList],
+function(D, semigroups, homomorphisms)
+  local efam, etype, type, maps, n, rtclosure, path, len, tobecomposed, gens,
+  out, s, i, j, paths, firsthom;
+
+  #  TODO would it be advisable to ensure all the semigroups are distinct?
```
Comment (wilfwilson) [inline gap/semigroups/semicons.gi:816]: I think it doesn't matter if there are repeated semigroups.

File: gap/semigroups/semicons.gi [hunk 6]
```diff
@@ -805,3 +805,227 @@ for _IsXSemigroup in ["IsTransformationSemigroup",
...
+                          "homomorphisms must commute");
+          fi;
+        od;
+        # If no errors so far, then all paths commute and we can add the comp.
+        Add(maps[i], firsthom);
+        # NB. perhaps a dynamic programming approach would be more
+        # efficient here.
+        # for some larger digraphs, the current method might compute some
+        # compositions of homomorphisms several times.
+        # This should be a FIXME.
```
Comment (wilfwilson) [inline gap/semigroups/semicons.gi:921]: Are you going to do this, then?

File: gap/semigroups/semicons.gi [hunk 7]
```diff
@@ -805,3 +805,227 @@ for _IsXSemigroup in ["IsTransformationSemigroup",
...
+      else
+        Add(maps[i], fail);
+        # is fail a reasonable thing to add here?
```
Comment (wilfwilson) [inline gap/semigroups/semicons.gi:924]: Yes, I think so.

File: gap/semigroups/semicons.gi (hunk 4 above)
Comment (wilfwilson) [reply gap/semigroups/semicons.gi:1019]: Okay, yeah remove this comment and then commented out code, then.

**Review by wilfwilson — CHANGES_REQUESTED** [review]
I've got some more comments, especially on the documentation. I've looked at the code now too (although not in precise detail), and it looks good.

**Comment by wilfwilson** [comment]
Also this PR now has conflicts with the `master` branch. I'd recommend rebasing and squashing all the commits into one big commit.

**Comment by tomcontileslie** [comment]
Thank you @wilfwilson for your comments so far. In order for me to address them faster, I'm migrating the work over to a new PR #722. I've made the most of the change to pull all the new commits from `semigroups:master`, and to squash down these commits into just one. I haven't yet addressed all of your review points but I've made changes for the first batch (sorry you won't be able to see them marked as outdated because of the migration). If you'd be keen to continue the review on #722 your comments are most welcome.

Perhaps @james-d-mitchell you could close this PR or change its labels?

File: doc/semicons.xml [hunk 8]
```diff
@@ -299,3 +299,174 @@ true]]></Example>
...
+      If <A>n</A> is a positive integer no greater than the number of vertices
+      in the underlying semilattice of <A>SSS</A>, and if <A>x</A> is an
+      element of the <A>n</A>th semigroup of <A>SSS</A>, then this function
+      returns an <Ref Attr="IsSSSE"/> object which is an element of <A>SSS</A>.
+      SSSEs of the same SSS can be compared and multiplied.
+      <P/>
+
+      If <C>el</C> is an SSSE, then the underlying strong semilattice of
+      semigroups object it belongs to can be recovered by calling
+      <C>StrongSemilatticeOfSemigroups(el)</C>.
```
[the remark this answers is not in the document]
Comment (tomcontileslie) [reply doc/semicons.xml:444]: Thank you Wilf. I have made all the changes suggested up to this point, and will work on your comments on the gap files at a later point (maybe next week).

File: gap/semigroups/semicons.gi (hunk 6 above)
Comment (tomcontileslie) [reply gap/semigroups/semicons.gi:921]: I think it involves dropping use of the `IteratorOfPaths` method and making everything a lot more complicated, and also threatens to take up lots of memory it not done correctly, so for now I've made it a proper FIXME and added more information on what the potential slowdown is. I may try to tackle this over winter but the initialisation still works as-is for now.

**Comment by james-d-mitchell** [comment]
Closing in favour of #722.

**Pull request closed without merging**